File: glox/internal/lox/token.go
package lox

import "fmt"

// Token represents group a characters with additional information that was
// obtained during the scanning phase.
type Token struct {
	Type    TokenType
	Lexeme  string
	Literal interface{}
	Line    int
}

// New creates a new token
func NewToken(typ TokenType, lexeme string, literal interface{}, line int) *Token {
	t := new(Token)
	t.Type = typ
	t.Lexeme = lexeme
	t.Literal = literal
	t.Line = line
	return t
}

func (t *Token) String() string {
	return fmt.Sprintf("%s %s %v", t.Type.String(), t.Lexeme, t.Literal)
}

var KeywordTokens = map[string]TokenType{
	"and":    AND,
	"class":  CLASS,
	"else":   ELSE,
	"false":  FALSE,
	"fun":    FUN,
	"for":    FOR,
	"if":     IF,
	"nil":    NIL,
	"or":     OR,
	"print":  PRINT,
	"return": RETURN,
	"super":  SUPER,
	"this":   THIS,
	"true":   TRUE,
	"var":    VAR,
	"while":  WHILE,
	"eof":    EOF,
}

/// TokenType is a just a wrapped string used to represent token's type
type TokenType uint

func (tt *TokenType) String() string {
	switch *tt {
	case L_PAREN:
		return "("
	case R_PAREN:
		return ")"
	case L_BRACE:
		return "{"
	case R_BRACE:
		return "}"
	case COMMA:
		return ","
	case DOT:
		return "."
	case MINUS:
		return "-"
	case PLUS:
		return "+"
	case SEMICOLON:
		return ";"
	case SLASH:
		return "/"
	case STAR:
		return "*"
	case BANG:
		return "!"
	case BANG_EQUAL:
		return "!="
	case EQUAL:
		return "="
	case EQUAL_EQUAL:
		return "=="
	case GREATER:
		return ">"
	case GREATER_EQUAL:
		return ">="
	case LESS:
		return "<"
	case LESS_EQUAL:
		return "<="
	case IDENT:
		return "IDENTIFIER"
	case STRING:
		return "STRING"
	case NUMBER:
		return "NUMBER"
	case AND:
		return "AND"
	case CLASS:
		return "CLASS"
	case ELSE:
		return "ELSE"
	case FALSE:
		return "FALSE"
	case FUN:
		return "FUN"
	case FOR:
		return "FOR"
	case IF:
		return "IF"
	case NIL:
		return "NIL"
	case OR:
		return "OR"
	case PRINT:
		return "PRINT"
	case RETURN:
		return "RETURN"
	case SUPER:
		return "SUPER"
	case THIS:
		return "THIS"
	case TRUE:
		return "TRUE"
	case VAR:
		return "VAR"
	case WHILE:
		return "WHILE"
	case EOF:
		return "EOF"
	}
	return ""
}

const (
	// Single-character tokens
	L_PAREN TokenType = iota
	R_PAREN
	L_BRACE
	R_BRACE
	COMMA
	DOT
	MINUS
	PLUS
	SEMICOLON
	SLASH
	STAR

	// One or two chracter tokens
	BANG
	BANG_EQUAL
	EQUAL
	EQUAL_EQUAL
	GREATER
	GREATER_EQUAL
	LESS
	LESS_EQUAL

	// Literals
	IDENT
	STRING
	NUMBER

	// Keywords
	AND
	CLASS
	ELSE
	FALSE
	FUN
	FOR
	IF
	NIL
	OR
	PRINT
	RETURN
	SUPER
	THIS
	TRUE
	VAR
	WHILE
	EOF
)

File: rlox/src/token.rs
/// Lox token. One difference of the implementation of glox is that we don't
/// eagerly evaluate the value of a literal.
#[derive(Debug)]
pub struct Token {
    /// Lox token type
    pub typ: Type,
    /// The string segment in source that represents this token.
    ///
    /// # Notes
    ///
    /// We are gonna copy the string segment that contains the lexem of
    /// our token, so we don't have to deal with rust's compile, this can be
    /// implemented more efficiently.
    pub lexeme: String,
    /// The position at which this token was found in source.
    pub pos: Position,
}

/// Lox token types
#[derive(Debug, PartialEq, Eq)]
pub enum Type {
    /// Single character '('
    LParen,
    /// Single character ')'
    RParen,
    /// Single character '{'
    LBrace,
    /// Single character '}'
    RBrace,
    /// Single character ','
    Comma,
    /// Single character '.'
    Dot,
    /// Single character '-'
    Minus,
    /// Single character '+'
    Plus,
    /// Single character ';'
    Semicolon,
    /// Single character '/'
    Slash,
    /// Single character '*'
    Star,
    /// Single character '!'
    Bang,
    /// Double character '!='
    BangEqual,
    /// Single character '='
    Equal,
    /// Double character '=='
    EqualEqual,
    /// Single character '>'
    Greater,
    /// Double character '>='
    GreaterEqual,
    /// Single character '<'
    Less,
    /// Double character '<='
    LessEqual,
    /// Named entity
    Ident,
    /// String literal
    String,
    /// Number literal
    Number,
    /// Keyword 'and'
    And,
    /// Keyword 'class'
    Class,
    /// Keyword 'else'
    Else,
    /// Boolean literal 'false'
    False,
    /// Keyword 'for'
    For,
    /// Keyword 'fun'
    Fun,
    /// Keyword 'if'
    If,
    /// Nothing literal 'nil'
    Nil,
    /// Keyword 'or'
    Or,
    /// Keyword 'print'
    Print,
    /// Keyword 'return'
    Return,
    /// Keyword 'super'
    Super,
    /// Keyword 'this'
    This,
    /// Boolean literal 'true'
    True,
    /// Keyword 'var'
    Var,
    /// Keyword 'while'
    While,
}

/// Position of the token in source
#[derive(Debug, Clone, Copy)]
pub struct Position {
    /// Current line in source file
    pub line: usize,
    /// Current column in source file
    pub column: usize,
}

impl Default for Position {
    fn default() -> Self {
        Self { line: 1, column: 1 }
    }
}

impl Position {
    /// Increment the line count by one and reset the column count
    pub fn next_line(&mut self) {
        self.line += 1;
        self.column = 1;
    }

    /// Increment the column count by one
    pub fn next_column(&mut self) {
        self.column += 1;
    }
}
